src/vcpkg/export.ifw.cpp

```diff
@@ -15,6 +15,86 @@ namespace vcpkg::Export::IFW
     using Dependencies::ExportPlanType;
     using Install::InstallDir;
 
+    std::string safe_rich_from_plain_text(StringView text)
+    {
+        // looking for `&`, not followed by:
+        // - '#<numbers>;`
+        // - '#x<hex numbers>;`
+        // - `<numbers, letters, or _>;`
+        // (basically, an HTML character entity reference)
+        using P = Parse::ParserBase;
+        constexpr static auto is_word = [](char ch) { return P::is_alphanum(ch) || ch == '_'; };
+        constexpr static auto is_hex_digit = [](char ch) {
+            return P::is_ascii_digit(ch) || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
+        };
+        constexpr static char escaped_amp[] = "&amp;";
+
+        auto first = text.begin();
+        auto last = text.end();
+
+        std::string result;
+        for (;;)
+        {
+            auto amp = std::find(first, last, '&');
+            result.append(first, amp);
+            first = amp;
+            if (first == last) break;
+
+            ++first; // skip amp
+            if (first == last)
+            {
+                result.append(escaped_amp);
+                break;
+            }
+            else if (*first == '#')
+            {
+                // &#123;
+                ++first; // skip hash
+                auto semi = std::find(first, last, ';');
+                bool is_char_ref = false;
+                if (semi != last && semi != first)
+                {
+                    if (*first == 'x' && semi != first + 1 && std::all_of(first + 1, semi, is_hex_digit))
+                    {
+                        is_char_ref = true;
+                    }
+                    else if (std::all_of(first, semi, P::is_ascii_digit))
+                    {
+                        is_char_ref = true;
+                    }
+                }
+
+                if (is_char_ref)
+                {
+                    result.append(amp, semi);
+                }
+                else
+                {
+                    result.append(escaped_amp);
+                    result.push_back('#');
+                    result.append(first, semi);
+                }
```

Comment:
```suggestion
                if (semi != last && semi != first
                  // hex character escape
                  && (*first == 'x' && semi != first + 1 && std::all_of(first + 1, semi, is_hex_digit)
                  // decimal character escape
                  || std::all_of(first, semi, P::is_ascii_digit)))
                {
                    // looks like a legitimate escape, so append it literally
                    first = amp;
                }
                else
                {
                    // doesn't look like a legitimate escape; so escape the &
                    result.append(escaped_amp);
                    result.push_back('#');
                }
                
                result.append(first, semi);
```


I am not a fan of introducing extra bools for control flow purposes and I think clarifying comments explaining why are better than the bool anyway. Note that this also code-dedupes the append.